### PR TITLE
feature: add cli attach command

### PR DESCRIPTION
**Pull request by ZYecho**
Signed-off-by: zhangyue <zy675793960@yeah.net>

### Ⅰ. Describe what this PR did
add attach cli command.

### Ⅱ. Does this pull request fix one issue?

None

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
Add proxy and term unit test.

### Ⅳ. Describe how to verify it
1. run a container and detach from terminal
`$ pouch run -it --detach-keys="ctrl-d" busybox`
2.  get id or name
`$ pouch ps`
3. attach it 
`$ pouch attach --detach-keys="ctrl-x" busybox`
### Ⅴ. Special notes for reviews
inspired by  [docker attach impl](https://github.com/docker/docker-ce/blob/8a582e89b407639018e067123eb17f61dbfd10ee/components/engine/pkg/term/ascii.go) and pr#2240

### CODE REVIEW

File: cli/attach.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+// runAttach is used to attach a container.
+func (ac *AttachCommand) runAttach(args []string) error {
+	name := args[0]
+
+	ctx := context.Background()
+	apiClient := ac.cli.Client()
+
+	in, out, err := setRawMode(!ac.noStdin, false)
+	if err != nil {
+		return fmt.Errorf("failed to set raw mode")
```
Comment (allencloud) [inline cli/attach.go]: please print the error as well @ZYecho

File: cli/attach.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	defer conn.Close()
+
+	var inReader io.Reader = os.Stdin
+	escapeKeys := term.DefaultEscapeKeys
+	// Wrap the input to detect detach escape sequence.
+	// Use default escape keys if an invalid sequence is given.
+	if ac.detachKeys != "" {
+		customEscapeKeys, err := term.ToBytes(ac.detachKeys)
+		if err != nil {
+			logrus.Warnf("invalid detach escape keys, using default: %s", err)
```
Comment (allencloud) [inline cli/attach.go]: Do we need to break here? 
Just confirm this.

File: cli/attach.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	<-wait
+	return nil
+}
+
+// example shows examples in attach command, and is used in auto-generated cli docs.
+func (ac *AttachCommand) example() string {
+	return `$ pouch ps
+Name     ID       Status    Image                              Runtime
+foo      71b9c1   Running   docker.io/library/busybox:latest   runc
+$ pouch attach foo`
```
Comment (allencloud) [inline cli/attach.go]: Will this command display the result?

**Comment by allencloud** [comment]
Do we need to add some integration test to cover this command line's functionality? @ZYecho

File: cli/attach.go (hunk 1 above)
Comment (ZYecho) [reply cli/attach.go]: Done

File: cli/attach.go (hunk 2 above)
Comment (ZYecho) [reply cli/attach.go]: Yeah, I saw docker break here, I'll keep this

File: cli/attach.go (hunk 3 above)
Comment (ZYecho) [reply cli/attach.go]: Sometimes the attach command don't get the result obviously, it depends on the how you use it.

**Comment by ZYecho** [comment]
> Do we need to add some integration test to cover this command line's functionality? @ZYecho

I did't find the pattern about how do the cli command integration test, is there is a doc or code sample about that? could you help me with this? thank you

**Comment by allencloud** [comment]
> the pattern about how do the cli command integration test, is there is a doc or code sample about that?

Yes, we have. Please check the documents here https://github.com/alibaba/pouch/blob/master/docs/test/test.md to get noticed of how to run local integration test. And an example is in https://github.com/alibaba/pouch/blob/master/test/cli_restart_test.go, you can see the restart command integration test there. @ZYecho 

Please feel free to tell us if you have any questions. Thanks.

**Comment by ZYecho** [comment]
> > the pattern about how do the cli command integration test, is there is a doc or code sample about that?
> 
> Yes, we have. Please check the documents here https://github.com/alibaba/pouch/blob/master/docs/test/test.md to get noticed of how to run local integration test. And an example is in https://github.com/alibaba/pouch/blob/master/test/cli_restart_test.go, you can see the restart command integration test there. @ZYecho
> 
> Please feel free to tell us if you have any questions. Thanks.

I add my integration test in test/cli_attach_test.go, but it seems not work. But when I test the attach command manually in the same way as code, it works correctly. Could you help to review the test code? Thank you.

File: cli/attach.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// runAttach is used to attach a container.
+func (ac *AttachCommand) runAttach(args []string) error {
+	name := args[0]
+
+	ctx := context.Background()
+	apiClient := ac.cli.Client()
+
+	in, out, err := setRawMode(!ac.noStdin, false)
```
Comment (fuweid) [inline cli/attach.go]: only terminal can support rawMode.

File: cli/attach.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	inReader = ioutils.NewReadCloserWrapper(term.NewEscapeProxy(os.Stdin, escapeKeys), os.Stdin.Close)
+
+	wait := make(chan struct{})
+	go func() {
+		io.Copy(os.Stdout, br)
+		wait <- struct{}{}
+	}()
+	go func() {
+		_, err := io.Copy(conn, inReader)
+		if _, ok := err.(term.EscapeError); ok {
```
Comment (fuweid) [inline cli/attach.go]: only copy when `no-stdin=false`

File: cli/attach.go (hunk 4 above)
Comment (fuweid) [inline cli/attach.go]: only terminal can support rawMode. we need to detect the command is running in the terminal.

File: pkg/term/proxy.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
+package term
```
Comment (fuweid) [inline pkg/term/proxy.go]: could we re-use `github.com/docker/docker/pkg/term` here?

File: cli/attach.go (hunk 5 above)
Comment (fuweid) [reply cli/attach.go]: we also need to close the conn after ending with conn.

File: pkg/term/proxy.go (hunk 6 above)
Comment (ZYecho) [reply pkg/term/proxy.go]: It's a good idea for pouch to depend on docker's package? I have no idea about it.

File: cli/attach.go (hunk 4 above)
Comment (ZYecho) [reply cli/attach.go]: Done

File: cli/attach.go (hunk 5 above)
Comment (ZYecho) [reply cli/attach.go]: updated

File: cli/attach.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+			fmt.Fprintf(os.Stderr, "failed to set raw mode %s", err)
+			return fmt.Errorf("failed to set raw mode")
+		}
+		defer func() {
+			if err := restoreMode(in, out); err != nil {
+				fmt.Fprintf(os.Stderr, "failed to restore term mode %s", err)
+			}
+		}()
+	} else {
+		return fmt.Errorf("can't use attach command in not terminal environment")
```
Comment (fuweid) [inline cli/attach.go]: why not in terminal?

File: pkg/term/proxy.go (hunk 6 above)
Comment (fuweid) [reply pkg/term/proxy.go]: but the content is from moby's package, right?

**Comment by ZYecho** [comment]
updated this pr, make integration test right , but CI blocked at criv1alpha2test, could you help to trigger the CI again?  @allencloud , ping @fuweid for review to move on this pr either.

**Comment by ZYecho** [comment]
@fuweid Maybe could reuse logDriver io backend instead of create another hijhack attach io backend to implement attach command? It's a good idea?

**Comment by ZYecho** [comment]
CI have passed, and fix some issue. @fuweid @allencloud PTAL, thanks.

File: vendor/vendor.json [hunk 8]
```diff
@@ -761,6 +761,12 @@
 			"revision": "fd2f2a919e392b96de74795ae9af2dc5e510bc4c",
 			"revisionTime": "2018-06-06T04:43:31Z"
 		},
+		{
+			"checksumSHA1": "GFsDxJkQz407/2nUBmWuafG+uF8=",
+			"path": "github.com/docker/docker/pkg/term",
+			"revision": "8e908cab46ee73104bedc4cf5fc4a15b7647b500",
+			"revisionTime": "2018-09-25T23:55:46Z"
```
Comment (fuweid) [inline vendor/vendor.json]: I think we should keep the same revision with existing docker package.

File: cli/attach.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+			fmt.Fprintf(os.Stderr, "failed to set raw mode %s", err)
+			return fmt.Errorf("failed to set raw mode")
+		}
+		defer func() {
+			if err := restoreMode(in, out); err != nil {
+				fmt.Fprintf(os.Stderr, "failed to restore term mode %s", err)
+			}
+		}()
+
+		escapeKeys := defaultEscapeKeys
```
Comment (fuweid) [inline cli/attach.go]: For the current design, we have not supported the `EscapeKeys` yet. Please check this.

File: cli/attach.go [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	conn, br, err := apiClient.ContainerAttach(ctx, name, !ac.noStdin)
+	if err != nil {
+		return fmt.Errorf("failed to attach container: %v", err)
+	}
+	defer conn.Close()
+
+	// double check in case container in wrong state
+	_, err = inspectAndCheckState(ctx, apiClient, name)
```
Comment (fuweid) [inline cli/attach.go]: I don't think we need to check it again. For current IO design, the pipe copy goroutine will exit if the task has been stopped.

File: vendor/vendor.json (hunk 8 above)
Comment (ZYecho) [reply vendor/vendor.json]: I have no idea how to keep it. could you give me some suggestions?

File: cli/attach.go (hunk 9 above)
Comment (ZYecho) [reply cli/attach.go]: just want to know, why the backend need work to support `EscapeKeys`?

File: cli/attach.go (hunk 10 above)
Comment (ZYecho) [reply cli/attach.go]: ok, update

**Comment by ZYecho** [comment]
close it temporarily

**Pull request closed without merging**